oak-api/src/main/java/org/apache/jackrabbit/oak/api/QueryEngine.java

```diff
@@ -91,6 +92,24 @@ Result executeQuery(
             String statement, String language, long limit, long offset,
             Map<String, ? extends PropertyValue> bindings,
             Map<String, String> mappings) throws ParseException;
+
+    /**
+     * Execute a query and get the result.
+     *
+     * @param statement the query statement
+     * @param language the language
+     * @param limit the maximum result set size (may not be negative but may be empty)
+     * @param offset the number of rows to skip (may not be negative but may be empty)
+     * @param bindings the bind variable value bindings
+     * @param mappings namespace prefix mappings
+     * @return the result
+     * @throws ParseException if the statement could not be parsed
+     * @throws IllegalArgumentException if there was an error executing the query
+     */
+    Result executeQuery(
+            String statement, String language, Optional<Long> limit, Optional<Long> offset,
```

Comment:
Some people suggest NOT using `Optional` as type parameters but only as return types. There are a lot of threads around this topic (see https://medium.com/@yassinhajaj/optionals-are-bad-practices-still-bad-practices-if-everyone-practices-them-6ec5a66c30aa).

IMHO it depends on the case. In this case, though, I would be in favor of removing it. The main reason is that `Optional` and `Long` are two wrappers. This double wrapping is hard to read and suboptimal for the compiler.

@thomasmueller I would like to hear your opinion on this

Reply:
@fabriziofortino 

I don't have a strong opinion about this.

As for the performance reason: for a method that is called a lot (like: put, log,...) , I would avoid Optional for performance reasons. But executeQuery isn't called that often, so that's not a good reason to avoid it.

I see some advantages of Optional over magic values (-1 for infinity: make sure you always treat the magic value correctly), and some disadvantages (a bit harder to read).

I would leave it up to the author, I think...

Reply:
@thomasmueller I agree. While it is double wrapping, it makes it clearer what the intent is. 

It also _seems_ like if you try to use a null Long Java will auto box the long -> Long resulting in the wrong method being called and failures executing queries. At least I think that's the issue. I tried doing this and running the tests and I _think_ that this is causing the tests to fail because it's calling the wrong method. As you said this can be handled through magic values but that makes it harder to evaluate if the user really intended (or messed up w/) Long.MAX_VALUE/-1 or just didn't set a value.

Reply:
Just a drive-by comment here, but instead of `Optional<Long>`, could you define a `Limit` type to serve as the wrapper for the limit long, and an `Offset` type to serve as the wrapper for the offset long? Then you would also eliminate the potential for callers accidentally reversing the order of the arguments.